assets/src/dashboard/app/api/apiProvider.js

```diff
@@ -118,19 +138,49 @@ export default function ApiProvider({ children }) {
           query,
         });
 
-        const serverStoryResponse = await apiFetch({
+        const response = await apiFetch({
           path,
+          parse: false,
         });
+
+        // TODO add headers for totals by status and have header reflect search
+        // only update totals when data is different
+        const totalStories = parseInt(response.headers.get('X-WP-Total'));
+        state.totalStories !== totalStories &&
+          dispatch({
+            type: STORY_ACTION_TYPES.UPDATE_TOTAL_STORIES_COUNT,
+            payload: totalStories,
+          });
+
+        const totalPages = parseInt(response.headers.get('X-WP-TotalPages'));
+        state.totalPages !== totalPages &&
+          dispatch({
+            type: STORY_ACTION_TYPES.UPDATE_TOTAL_STORIES_PAGES,
+            payload: totalPages,
+          });
+
+        const serverStoryResponse = await response.json();
+
         const reshapedStories = serverStoryResponse
           .map(reshapeStoryObject(editStoryURL))
           .filter(Boolean);
-        setStories(reshapedStories);
+
+        dispatch({
+          type: STORY_ACTION_TYPES.UPDATE_STORIES,
+          payload: reshapedStories,
+        });
```

Comment:
We have three dispatch calls that could potentially force 3 renders. `useReducer` unlike Redux with Thunk middleware will not batch these, but fire them sequentially. It would be best to create a `FETCH_STORIES_SUCCESS` or `FETCH_STORIES_FAILURE` action with the entire payload that updates the state that includes count, total pages, and the server story response where appropriate.

Reply:
great call out, thank you! i was treating useReducer like a redux reducer. I've updated so that it'll batch update and i also moved the check for if all pages have been fetched to the reducer so that we can calculate that all together. In the same way, am checking to see if we're loading page 1, if we're on page 1 we should start over with our sort which decides what gets viewed. No more clearing state separately. 

I think this reduces renders and is getting more straight forward. 

I'm hesitant to add in the 'isLoading' state right now, i don't think it's really needed. We have a loading message until there's nothing else to load in pagination and ntohing on the page is really waiting for a loading boolean. 

I did go ahead and add in FETCH_STORIES_FAILURE with just a bool, i'm not doing anything with it - figure error handling is it's own thing.